File: backend/app/routers/repos.py
from fastapi import APIRouter
from app.schemas.repo_schema import RepoRequest, RepoIssuesResponse
from app.services import github_service

router = APIRouter(prefix="/api/repos", tags=["repos"])


@router.post("/analyze", response_model=RepoIssuesResponse)
async def analyze_repo(payload: RepoRequest):
    owner, name = github_service.parse_repo_url(payload.repo_url)

    metadata = await github_service.fetch_repo_metadata(owner, name)
    issues_data = await github_service.fetch_repo_issues(
        owner,
        name,
        state="open",
        max_issues=payload.per_page,
        page=payload.page,
        search=payload.search,
        sort=payload.sort,
    )

    return RepoIssuesResponse(
        owner=owner,
        name=name,
        total_open_issues=metadata["open_issues_count"],
        issues=issues_data["issues"],
        page=issues_data["page"],
        per_page=issues_data["per_page"],
        has_next=issues_data["has_next"],
        has_previous=issues_data["has_previous"],
        search=payload.search,
        sort=payload.sort,
        analytics_note=f"Analytics are calculated from the currently loaded page of {len(issues_data['issues'])} issues.",
    )

File: frontend/src/api/repos.ts
import axios from "axios";

const API_BASE_URL = import.meta.env.VITE_API_URL || "http://localhost:8000";

export const apiClient = axios.create({
  baseURL: API_BASE_URL,
  timeout: 20000,
});

export interface Issue {
  id: number;
  number: number;
  title: string;
  body: string | null;
  state: string;
  labels: string[];
  comments: number;
  author: string | null;
  created_at: string;
  updated_at: string;
  html_url: string;
}

export type IssueSort = "recent" | "most_commented" | "oldest";

export interface RepoIssuesResponse {
  owner: string;
  name: string;
  total_open_issues: number;
  issues: Issue[];
  page: number;
  per_page: number;
  has_next: boolean;
  has_previous: boolean;
  search: string;
  sort: IssueSort;
  analytics_note: string;
}

export interface AnalyzeRepoOptions {
  page?: number;
  perPage?: number;
  search?: string;
  sort?: IssueSort;
}

export async function analyzeRepo(
  repoUrl: string,
  options: AnalyzeRepoOptions = {}
): Promise<RepoIssuesResponse> {
  const response = await apiClient.post<RepoIssuesResponse>("/api/repos/analyze", {
    repo_url: repoUrl,
    page: options.page ?? 1,
    per_page: options.perPage ?? 100,
    search: options.search ?? "",
    sort: options.sort ?? "recent",
  });
  return response.data;
}
